### ISSUE TITLE
Possible retain cycle

### ISSUE THREAD

**Issue by chaserCN** [ReactiveKit/ReactiveKit#14]
Hi guys. What i was trying to do is to create a custom operator that could restart Operation after some delay. So i took 'retry' function and changed it a bit. But when i started testing, it looked like i had some memory leaks or retain cycles. I returned back to ‘retry’  but got same results.

Here is what i have.

A simple downloader class which always generates errors in order to test ‘retry’.
```swift
class Service {
   static func operationToTestRetainCycle(address address: String) -> Operation<Int, NSError> {
        return Operation { observer in
            
            print("[Service]: send request")
            
            let request = Alamofire
                .request(.GET, address)
                .responseString { response in
                    let error = NSError(domain: "TestDomain", code: 0, userInfo: nil)
                    observer.failure(error)
                }
            
            return BlockDisposable() {
                request.cancel()
            }
        }
    }
}
```
A worker class. An instance created after a tap on a button and destroyed after another tap.

```swift
private class Worker {
    var memTest = MemTest(name: "Worker")
    var operation: Operation<Int, NSError>?
    var disposable: DisposableType?
    
    deinit {
        disposable?.dispose()
    }
    
    func start() {
        operation = Service.operationToTestRetainCycle(address: "http://google.com").retry(3)
        disposable = operation!.observe { event in
            NSLog("[Worker] received event \(event)")
        }
    }
}
```
I wanted to check that all Operations are released after the Worker object is released, but i could not add deinit, because Operation is a structure. So i added a helper class MemTest:

```swift
public class MemTest {
    let name: String
    
    public init(name: String) {
        self.name = name
        NSLog("[\(name)]: init \(unsafeAddressOf(self))")
    }
    
    deinit {
        NSLog("[\(self.name)]: deinit \(unsafeAddressOf(self))")
    }
}
```
And added a variable to the Operation:

```swift
public struct Operation<Value, Error: ErrorType>: OperationType {
  private let memTest = MemTest(name: "Operation")
…
}
```
When Operation is released, MemTest is released and i can see logs.

Logs are:
```
<Tap>
[Worker]: init 0x00007fdab242bcd0
[Operation]: init 0x00007fdab270b960
[Operation]: init 0x00007fdab2714ab0
[Service]: send request
[Service]: send request
[Service]: send request
[Service]: send request
[Worker] received event Failure(Error Domain=TestDomain Code=0 "(null)")
<Tap>
[Worker]: deinit 0x00007fdab242bcd0
[Operation]: deinit 0x00007fdab2714ab0
```
As you can see MemTest with address 0x00007fdab270b960 is not released.

Trying to fix i transformed Operation into a class and added [weak self] in the ‘retry’:
```swift
    @warn_unused_result
  public func retry(var count: Int) -> Operation<Value, Error> {
    return create { observer in
      let serialDisposable = SerialDisposable(otherDisposable: nil)
      
      var attempt: (() -> Void)!
      
      attempt = {[weak self] in
        serialDisposable.otherDisposable?.dispose()
        serialDisposable.otherDisposable = self?.observe(on: nil) { event in
          switch event {
          case .Failure(let error):
            if count > 0 {
              count--
              attempt()
            } else {
              observer.failure(error)
            }
          default:
            observer.observer(event._unbox)
          }
        }
      }
      
      attempt()
      return serialDisposable
    }
  }
```
Logs after:
```
<Tap>
[Worker]: init 0x00007fdc2b526f30
[Operation]: init 0x00007fdc2b54da60
[Operation]: init 0x00007fdc2b540370
[Service]: send request
[Service]: send request
[Service]: send request
[Service]: send request
[Worker] received event Failure(Error Domain=TestDomain Code=0 "(null)")
<Tap>
[Worker]: deinit 0x00007fdc2b526f30
[Operation]: deinit 0x00007fdc2b540370
[Operation]: deinit 0x00007fdc2b54da60
```

**Comment by chaserCN** [comment]
Btw, if I create a MemTest variable and make 'attempt' closure to capture it, the MemTest is never released, though both operations are:

```swift
 @warn_unused_result
  public func retry(var count: Int) -> Operation<Value, Error> {
    return create { observer in
      let serialDisposable = SerialDisposable(otherDisposable: nil)

        let retryProducerMemTest = MemTest(name: "producer") // NEW CODE
        
      var attempt: (() -> Void)!
        
      attempt = {[weak self] in // NEW CODE
        
        print("attempt is called, \(retryProducerMemTest.name)") // NEW CODE
        
        serialDisposable.otherDisposable?.dispose()
        serialDisposable.otherDisposable = self?.observe(on: nil) { event in
          switch event {
          case .Failure(let error):
            if count > 0 {
              count--
              attempt()
            } else {
              observer.failure(error)
            }
          default:
            observer.observer(event._unbox)
          }
        }
      }
      
      attempt()
      return serialDisposable
    }
  }
```

**Comment by srdanrasic** [comment]
Great catch @chaserCN and thanks for the detailed analysis!

It appears that the `attempt` is causing leaks. I think it should be enough just to nullify it on disposing. Something like this:

```swift
  @warn_unused_result
  public func retry(var count: Int) -> Operation<Value, Error> {
    return create { observer in
      let serialDisposable = SerialDisposable(otherDisposable: nil)
      
      var attempt: (() -> Void)?
      
      attempt = {
        serialDisposable.otherDisposable?.dispose()
        serialDisposable.otherDisposable = self.observe(on: ImmediateExecutionContext) { event in
          switch event {
          case .Failure(let error):
            if count > 0 {
              count--
              attempt?()
            } else {
              observer.failure(error)
            }
          default:
            observer.observer(event._unbox)
          }
        }
      }
      
      attempt?()

      return BlockDisposable {
        serialDisposable.dispose()
        attempt = nil
      }
    }
  }
```

I'll try to confirm tomorrow.

**Comment by srdanrasic** [comment]
Should be fixed in v1.0.9.
